Update README to have more concise instruction and usage for Neo4j ex… (#10)
Bit-Quill/amazon-neptune-tools: neo4j-to-neptune/readme.md

### Usage

  - [`convert-csv`](docs/convert-csv.md)

## Migration Process

> Migration of data from Neo4j to Neptune can now be accomplished in a few ways:

Reviewer: ```suggestion
Migration of data from Neo4j to Neptune can be accomplished in a few ways:
```

I've always considered Readme's to be "timeless". So the "now" in that sentence felt a bit off to me. But that is strictly an opinion.

Author: i agree with the README should be "timeless"

> ### Option 1: Manual Multi-Step Process

Reviewer: ```suggestion
### Option 1: Manual Multi-Step
```
I think these read better without "Process" on each one.

 1. [**Export CSV from Neo4j**](#export-csv-from-neo4j) – Use the APOC export procedures to export data from Neo4j to CSV.
 2. [**Convert CSV**](#convert-csv) – Use the `convert-csv` command-line utility to convert the exported CSV into the Neptune Gremlin bulk load CSV format.
 3. [**Bulk Load into Neptune**](#bulk-load-into-neptune-manual) – Use the Neptune bulk load API to load data into Neptune.

> ### Option 2: Local Convert and Bulk Load Process

Reviewer: ```suggestion
### Option 2: Local Convert then Bulk Load
```
I think these read better without "Process" on each one. Changed to "then" to be consistent with the next option.

 1. [**Export CSV from Neo4j**](#export-csv-from-neo4j) – Use the APOC export procedures to export data from Neo4j to CSV.
 2. [**Convert and Bulk Load**](#convert-and-bulk-load) – Use the `convert-csv` command with either `--bulk-load-config` or the required combination of `--bucket-name`, `--neptune-endpoint`, and `--iam-role-arn` to automatically convert the CSV, upload it to S3, and initiate a bulk load into Neptune.

> ### Option 3: Automatically Stream Convert and Bulk Load Process

Reviewer: ```suggestion
### Option 3: Automatically Stream and Convert then Bulk Load
```
I think these read better without "Process" on each one.

Also, this one reads a bit weird with "Stream Convert" having me scratch my head in confusion. So I massaged it a bit more.

#### 2. Enable exports

> Create/Update the _apoc.conf_ configuration file to enable exports:

Reviewer: ```suggestion
Create or update the `apoc.conf` configuration file to enable exports:
```

```
CALL apoc.export.csv.all(
 "neo4j-export.csv",
 {d:','}
)
```

**Note:** When running this command please use the syntax above, do not include `{stream:true}` in the command.  Streaming the results back to the browser and then downloading them as a CSV will result in a file that will not be correctly processed by the conversion utility.

The path that you specify for the export file will be resolved relative to the Neo4j _import_ directory. `apoc.export.csv.all` creates a single CSV file containing data for all nodes and relationships.

### Convert CSV

Use the [`convert-csv`](docs/convert-csv.md) command-line utility to convert the CSV files exported from Neo4j into the Neptune Gremlin bulk load CSV format.

The utility requires two sets of parameters:

> 1. Input from Neo4j — Provide one of the following:

Reviewer: ```suggestion
1. **Input method from Neo4j** — Provide one of the following:
```

>   - [Manual] A path to locally exported Neo4j CSV file(s) using the `--input` or `-i` flag.

Reviewer: Since you have "file(s)" it implies you can have more than one. It might be nice to mention how multiple files are specified or link to the spot in the docs that describes the process in full.

Author: This should really just be `files` since the export from Neo4j will always result in 2 files as of right now. `edges.csv` and `vertices.csv`

Author: changed it to just `files`

  - [Stream] A path to a `.env` file using the `--dotenv-file` or `-df` flag. This file must define the variables `NEO4J_URI`, `NEO4J_USERNAME`, and `NEO4J_PASSWORD`, corresponding to the Neo4j URI, username, and password.
  - [Stream] Directly specify the Neo4j URI, username, and password using the flags `--uri` or `-u`, `--username` or `-n`, and `--password` or `-pw`

> 2. Output directory — The path to the local directory where the converted CSV files will be written.

Reviewer: ```suggestion
2. **Output directory** — The path to the local directory where the converted CSV files will be written.
```

Note that `convert-csv` will always use a __double__ for values with decimal or scientific notation.

#### Configuration Methods

##### Method 1: Using Local Exported Neo4j CSV

> To convert a local csv that was exported from Neo4j, provide the required parameters:

Reviewer: ```suggestion
To convert a local CSV that was exported from Neo4j, provide the required parameters:
```

```bash
java -jar neo4j-to-neptune.jar convert-csv \
  --input /path/to/neo4j-export.csv \
  --dir /path/to/output \
  [additional options]
```

##### Method 2: Using Stream with Environment files

To stream data from Neo4j to Neptune

```bash
java -jar neo4j-to-neptune.jar convert-csv \
  --dotenv-file /path/to/dotenv/file \
  --dir /path/to/output \
  [additional options]
```

##### Method 3: Using Stream with Neo4j credentials files

To stream data from Neo4j to Neptune

```bash
java -jar neo4j-to-neptune.jar convert-csv \
  --uri <Neo4j connection uri> \
  --username <Neo4j username> \
  --password <Neo4j password> \
  --dir /path/to/output \
  [additional options]
```

#### Configuration with YAML

> The `convert-csv` utility supports conversion configuration through the YAML file for ID transformation, label mapping, and filtering:

Reviewer: ```suggestion
The `convert-csv` utility conversion process can be configured through the YAML file for ID transformation, label mapping, and filtering:
```

```bash
java -jar neo4j-to-neptune.jar convert-csv \
  --input /path/to/neo4j-export.csv \
  --dir /path/to/output \
  --conversion-config conversion-config.yaml \
  [additional options]
```

> **Example configuration file:**

Reviewer: ```suggestion
##### Example configuration file
```

Reviewer: This might also work well as a note callout.

```
> [!NOTE]
> For an example of the conversion config YAML file refer to [`docs/example-conversion-config.yaml`](docs/example-conversion-config.yaml)
```

> For an example of the conversion config YAML file refer to `docs/example-conversion-config.yaml`

Reviewer: ```suggestion
For an example of the conversion config YAML file refer to [`docs/example-conversion-config.yaml`](docs/example-conversion-config.yaml)
```

> **ID Transformation Templates:**

Reviewer: ```suggestion
##### ID Transformation Templates
```

Templates can reference original data fields using placeholders:

- **Vertex templates**: `{_id}`, `{_labels}`, `{property_name}`
- **Edge templates**: `{_type}`, `{_start}`, `{_end}`, `{~from}`, `{~to}`, `{property_name}`

> **Examples:**

Reviewer: ```suggestion
##### Examples
```

- `"{_labels}_{name}_{_id}"` → `"Person_John_123"`
- `"e!{~label}_{~from}_{~to}"` → `"e!KNOWS_Person_123_Person_456"`

The utility uses two-pass processing: first transforming vertices and building ID mappings, then processing edges with correct vertex references. When vertices are skipped, connected edges are automatically skipped to maintain graph consistency.

### Bulk Load into Neptune

Use the [Neptune bulk loader](https://docs.aws.amazon.com/neptune/latest/userguide/bulk-load.html) to load data into Neptune from the converted CSV files.

> ### Convert and Bulk Load

Reviewer: ```suggestion
### Convert then Bulk Load
```

1. **AWS Credentials**: Configure AWS credentials using one of the following methods:
   - AWS CLI: `aws configure`
   - Environment variables: `AWS_ACCESS_KEY_ID` and `AWS_SECRET_ACCESS_KEY`
   - IAM roles (if running on EC2)
   - AWS credentials file

2. **S3 Bucket**: An S3 bucket where CSV files will be uploaded, where the region is the same as your Neptune Cluster

3. **IAM Role**: An IAM role with permissions to:
   - Read from the S3 bucket
   - Access Neptune cluster
   - Perform bulk load operations

4. **Neptune Cluster**: A running Neptune cluster with bulk load enabled, where the region is the same as your S3 bucket

#### Configuration Methods

##### Method 1: Using CLI Parameters

To enable bulk load using CLI parameters, provide the required Neptune configuration:

```bash
java -jar neo4j-to-neptune.jar convert-csv \
  --input /path/to/neo4j-export.csv \
  --dir /path/to/output \
  --neptune-endpoint your-cluster.cluster-abc123.region.neptune.amazonaws.com \
  --bucket-name your-s3-bucket \
  --iam-role-arn arn:aws:iam::account:role/YourNeptuneRole \
  [additional options]
```

##### Method 2: Using Configuration File

> Create a YAML configuration file with your bulk load settings:
> 
> **bulk-load-config.yaml:**

Reviewer: ```suggestion
Create a YAML configuration file named `bulk-load-config.yaml` with your bulk load settings:

```

```yaml
# S3 Configuration
bucket-name: "my-neptune-bulk-load-bucket"
s3-prefix: "neptune-data"

# Neptune Configuration
neptune-endpoint: "my-neptune-cluster.cluster-abc123def456.us-east-1.neptune.amazonaws.com"

# IAM Configuration
iam-role-arn: "arn:aws:iam::123456789012:role/NeptuneLoadFromS3Role"

# Performance Settings
parallelism: "OVERSUBSCRIBE"  # Options: LOW, MEDIUM, HIGH, OVERSUBSCRIBE

# Monitoring
monitor: true
```

Then use the configuration file:

```bash
java -jar neo4j-to-neptune.jar convert-csv \
  --input /path/to/neo4j-export.csv \
  --dir /path/to/output \
  --bulk-load-config bulk-load-config.yaml \
  [additional options]
```

##### Method 3: Hybrid Approach (Configuration File + CLI Overrides)

You can use a configuration file and override specific parameters via CLI:

```bash
java -jar neo4j-to-neptune.jar convert-csv \
  --input /path/to/neo4j-export.csv \
  --dir /path/to/output \
  --bulk-load-config bulk-load-config.yaml \
  --neptune-endpoint different-cluster.cluster-xyz789.us-west-2.neptune.amazonaws.com \
  --parallelism HIGH \
  [additional options]
```

> **Parameter Precedence:** CLI parameters override configuration file values, which override default values.

Reviewer: ```suggestion
> [!IMPORTANT]
> **Parameter Precedence:** CLI parameters override configuration file values, which override default values.
```
